### PR TITLE
Mvp

### PR DESCRIPTION
**Pull request by mul53**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,12 +4,13 @@
   "description": "node-skeleton",
   "main": "app.js",
   "scripts": {
-    "start": "nodemon --trace-warnings ./src/app.js",
+    "start": "nodemon ./src/app.ts",
     "debug": "node --inspect-brk --trace-warnings ./src/app.js",
-    "start-prod": "node ./src/app.js",
+    "start-prod": "NODE_ENV=production ts-node-transpile-only ./src/server.ts",
```
Comment (leonprou) [inline package.json:9]: `NODE_ENV=production` is required right?
Should it be added to the node-skeleton?

File: src/services/event.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    get provider () {
+      return this.providerService.getProvider()
+    }
+
+    addEvents () {
+      this.addWalletTransferToEvent()
+    }
+
+    addWalletTransferToEvent () {
```
Comment (leonprou) [inline src/services/event.ts]: can you add another handler that will monitor native Fuse transactions?

File: src/controllers/api/v1/subscriptions/index.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
+import { Request, Response, NextFunction } from 'express'
+import Container from 'typedi'
+import SubscriptionService from '../../../../services/subscription'
```
Comment (leonprou) [inline src/controllers/api/v1/subscriptions/index.ts]: can you use `moduleAliases` for this? writing from '@services/serviceName`

File: src/controllers/api/v1/subscriptions/index.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      )
+
+      subscriptionService.unsubscribe(
+        WALLET_TRANSFER_TO_EVENT,
+        address
+      )
+
+      res.json({ message: 'Successfully unsubscribed from event' })
+    } catch (e) {
+      next(e)
```
Comment (leonprou) [inline src/controllers/api/v1/subscriptions/index.ts:59]: What happends when you use `next` with exception?
Usually we do not catch such errors, but have an error handler per application

File: src/services/event.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      this.eventManager.on(WALLET_TRANSFER_TO_EVENT, async (data, to) => {
+        try {
+          const subscription: any = await this.userSubscriptionService
+            .getSubscription(
+              WALLET_TRANSFER_TO_EVENT,
+              to
+            )
+
+          if (subscription && subscription.webhookUrl) {
+            await axios.post(subscription.webhookUrl, data)
```
Comment (leonprou) [inline src/services/event.ts]: I think you can remove the await here

File: src/services/subscription.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import { Service } from 'typedi'
+
+@Service()
+export default class SubscriptionService {
+    private readonly subscriptions: Map<string, Array<string>>
+
+    constructor () {
+      this.subscriptions = new Map()
```
Comment (leonprou) [inline src/services/subscription.ts]: Can you move this map to `redis`? That way the storage will be both efficient and persistant

File: src/routes/api/v1/index.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import SubscriptionsController from '../../../controllers/api/v1/subscriptions'
+import {
+  subscribeWalletTransfersToValidation,
+  unsubscribeWalletTransfersToValidation
+} from '../../../controllers/api/v1/subscriptions/validations'
+
+const router = Router()
+
+router.post(
+  '/subscribe/wallet/transfers-to',
```
Comment (leonprou) [inline src/routes/api/v1/index.ts]: IMO `/wallet` part of the route is not needed. We can keep it as:
`subscribe/:eventName`, like `subscribe/transfers-to`

File: src/controllers/api/v1/subscriptions/index.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        webhookUrl
+      } = req.body
+
+      const errors = validationResult(req)
+      if (!errors.isEmpty()) {
+        res.status(400).json({ error: errors.mapped() })
+        return
+      }
+
+      const userService = Container.get(UserService)
```
Comment (leonprou) [inline src/controllers/api/v1/subscriptions/index.ts:13]: Can you think of different names for this 3 services? They all names kinda same though functionality differ.
IMO at least one of them need to be changed to make it more clear

**Review by leonprou — CHANGES_REQUESTED** [review]
Looks good!
Some small comments added.
We also would like to add `redis` to save the subscriptions map

File: package.json (hunk 1 above)
Comment (mul53) [reply package.json:9]: yes, i can add that to the skeleton

File: src/services/event.ts (hunk 2 above)
Comment (mul53) [reply src/services/event.ts]: Since this is on hold, I will resolve this

**Review by leonprou — APPROVED** [review]

**Pull request merged**